File: cpp-client/deephaven/client/src/chunk/chunk_maker.cc
#include "deephaven/client/chunk/chunk_maker.h"

#include "deephaven/client/column/column_source.h"

using deephaven::client::column::ColumnSourceVisitor;
using deephaven::client::column::DoubleColumnSource;
using deephaven::client::column::Int32ColumnSource;
using deephaven::client::column::Int64ColumnSource;
using deephaven::client::column::UInt64ColumnSource;

namespace deephaven::client::chunk {
namespace {
struct Visitor final : ColumnSourceVisitor {
  explicit Visitor(size_t chunkSize) : chunkSize_(chunkSize) {}

  void visit(const Int32ColumnSource &source) final {
    result_ = Int32Chunk::create(chunkSize_);
  }

  void visit(const Int64ColumnSource &source) final {
    result_ = Int64Chunk::create(chunkSize_);
  }

  void visit(const UInt64ColumnSource &source) final {
    result_ = UInt64Chunk::create(chunkSize_);
  }

  void visit(const DoubleColumnSource &source) final {
    result_ = DoubleChunk::create(chunkSize_);
  }

  size_t chunkSize_;
  AnyChunk result_;
};
}  // namespace

AnyChunk ChunkMaker::createChunkFor(const ColumnSource &columnSource,
    size_t chunkSize) {
  Visitor v(chunkSize);
  columnSource.acceptVisitor(&v);
  return std::move(v.result_);
}
}  // namespace deephaven::client::chunk

File: csharp/client/ManagedClient/chunk/ChunkMaker.cs
namespace Deephaven.ManagedClient;

public static class ChunkMaker {
  public static Chunk CreateChunkFor(IColumnSource columnSource, int chunkSize) {
    var visitor = new ChunkMakerVisitor(chunkSize);
    columnSource.Accept(visitor);
    return visitor.Result!;
  }

  private class ChunkMakerVisitor(int chunkSize) :
    IColumnSourceVisitor<ICharColumnSource>,
    IColumnSourceVisitor<IByteColumnSource>,
    IColumnSourceVisitor<IInt16ColumnSource>,
    IColumnSourceVisitor<IInt32ColumnSource>,
    IColumnSourceVisitor<IInt64ColumnSource>,
    IColumnSourceVisitor<IFloatColumnSource>,
    IColumnSourceVisitor<IDoubleColumnSource>,
    IColumnSourceVisitor<IBooleanColumnSource>,
    IColumnSourceVisitor<IStringColumnSource>,
    IColumnSourceVisitor<ITimestampColumnSource>,
    IColumnSourceVisitor<ILocalDateColumnSource>,
    IColumnSourceVisitor<ILocalTimeColumnSource> {
    public Chunk? Result { get; private set; }

    public void Visit(ICharColumnSource cs) {
      Result = CharChunk.Create(chunkSize);
    }

    public void Visit(IByteColumnSource cs) {
      Result = ByteChunk.Create(chunkSize);
    }

    public void Visit(IInt16ColumnSource cs) {
      Result = Int16Chunk.Create(chunkSize);
    }

    public void Visit(IInt32ColumnSource cs) {
      Result = Int32Chunk.Create(chunkSize);
    }

    public void Visit(IInt64ColumnSource cs) {
      Result = Int64Chunk.Create(chunkSize);
    }

    public void Visit(IFloatColumnSource cs) {
      Result = FloatChunk.Create(chunkSize);
    }

    public void Visit(IDoubleColumnSource cs) {
      Result = DoubleChunk.Create(chunkSize);
    }

    public void Visit(IBooleanColumnSource cs) {
      Result = BooleanChunk.Create(chunkSize);
    }

    public void Visit(IStringColumnSource cs) {
      Result = StringChunk.Create(chunkSize);
    }

    public void Visit(ITimestampColumnSource cs) {
      Result = DhDateTimeChunk.Create(chunkSize);
    }

    public void Visit(ILocalDateColumnSource cs) {
      Result = LocalDateChunk.Create(chunkSize);
    }

    public void Visit(ILocalTimeColumnSource cs) {
      Result = LocalTimeChunk.Create(chunkSize);
    }

    public void Visit(IColumnSource cs) {
      throw new Exception($"Programming error: No visitor for type {cs.GetType().Name}");
    }
  }
}
